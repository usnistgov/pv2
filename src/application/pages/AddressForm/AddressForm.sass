.address-form-container
  padding-left: 64px
  padding-right: 64px
  padding-bottom: 64px

.address-form-text
  text-align: center
  margin: 32px 16px
  padding: 0 64px

.map
  padding: 0
  height: 100%

  iframe
    width: 100%
    height: 100%