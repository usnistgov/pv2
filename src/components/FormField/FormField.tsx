import React, {ReactElement, useState} from "react";

// Library Imports
import {InputAdornment, TextField, Tooltip} from "@material-ui/core";
import {InputProps as StandardInputProps} from "@material-ui/core/Input/Input";
import {mdiInformation} from "@mdi/js";
import {Icon as MdiIcon} from "@mdi/react";

// User Imports
import {ReduxGetSet} from "../../Utils";

// Stylesheets
import "../Info.sass";

export interface FormFieldProps<T> {
    // Label to display in TextField.
    label: string;

    // Yup schema to validate input against.
    schema: any;

    // The redux value to get and set.
    value: ReduxGetSet<T>;

    // Displays required asterisk if true.
    required?: boolean;

    // The type for the input field. Default is text.
    type?: string;

    // Text to display at the beginning of the input field.
    startAdornment?: string;

    // Text ot display at the end of the input field.
    endAdornment?: string;

    // Text to display in tooltip
    tooltip?: string;

    // Information icon text
    info?: string;
}

/*
 * FormField is a convenience wrapper around the Material-UI TextField. It handles
 * input validation and errors dynamically as the user types, and assigns input
 * directly to the provided redux store object if valid.
 */
export default function FormField<T>(props: FormFieldProps<T>): ReactElement {
    // Error state
    const [hasError, setHasError] = useState(false);
    const [errorMessages, setErrorMessages] = useState<string[]>([]);

    /*
     * Returns the start or end adornments based on the props.
     */
    function getInputProps(): StandardInputProps {
        let inputProps: StandardInputProps = {};

        if (props.startAdornment !== undefined) {
            inputProps['startAdornment'] = <InputAdornment position={"start"}>{props.startAdornment}</InputAdornment>
        }
        if (props.endAdornment !== undefined) {
            inputProps['endAdornment'] = <InputAdornment position={"end"}>{props.endAdornment}</InputAdornment>
        }

        return inputProps;
    }

    const field = (
        <TextField fullWidth
                   type={props.type}
                   required={props.required}
                   error={hasError}
                   helperText={errorMessages}
                   label={props.label}
                   defaultValue={props.value.get()}
                   variant={"filled"}
                   onChange={(event) => {
                       props.schema.validate(event.target.value)
                           .finally(() => {
                               props.value.set(props.schema.cast(event.target.value));
                               setHasError(false);
                               setErrorMessages([]);
                           })
                           .catch((schemaError: { errors: any; }) => {
                               setHasError(true);
                               setErrorMessages(schemaError.errors);
                           });
                   }}
                   InputProps={getInputProps()}
        />
    );

    const withTooltip = props.tooltip ? <Tooltip title={props.tooltip}>{field}</Tooltip> : field;

    return (props.info ?
            <div className={"with-icon"}>
                {withTooltip}
                <Tooltip title={props.info}>
                    <MdiIcon className={"icon"} path={mdiInformation} size={1.2} color={'#898989'}/>
                </Tooltip>
            </div>
            : withTooltip
    );
}
