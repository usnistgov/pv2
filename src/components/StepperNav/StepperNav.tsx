import {Children, ReactElement, useContext, useState} from 'react';

// Library Imports
import {Button, Grid, Step, StepLabel, Stepper} from "@material-ui/core";
import {Icon as MdiIcon} from "@mdi/react"
import {mdiArrowLeft, mdiArrowRight, mdiCheck, mdiClose} from "@mdi/js";
import {useHistory} from "react-router-dom";

// Stylesheets
import "./StepperNav.sass"
import {observer} from "mobx-react-lite";
import {Store} from "../../application/ApplicationStore";
import {autorun} from "mobx";

export interface StepperNavProps {
    // List of StepperPage components that should be a part of the Stepper Nav.
    children: ReactElement[];

    // Event handler for when finish button is clicked.
    onFinish?: () => void;
}

/*
 * Creates a navigation component that steps between the provided list of child pages with
 * a next and back button.
 */
const StepperNav = observer(({children, onFinish}: StepperNavProps) => {
    const store = useContext(Store).formUiStore;

    // React-router history object
    const history = useHistory();

    // Boolean values for view generation
    const isFirstStep: boolean = store.current <= 0;
    const isLastStep: boolean = store.current >= children.length - 1;

    /*
     * Creates a label from the StepperPage component that is passed in.
     * This function should only be called with a StepperPage component as the argument.
     */
    function createStepLabel(page: any): ReactElement {
        return (
            <Step key={page.props.label}>
                <StepLabel>{page.props.label}</StepLabel>
            </Step>
        );
    }

    return (
        <>
            <Grid className={"grid-container"} container justify={'center'} alignItems={'center'} spacing={0}>
                <Grid item xs={1}>
                    <Button
                        onClick={isFirstStep ? history.goBack : () => store.previous()}
                        startIcon={
                            <MdiIcon path={isFirstStep ? mdiClose : mdiArrowLeft}
                                     size={1}/>
                        }
                        data-testid={"back-button"}
                    >
                        {isFirstStep ? "Cancel" : "Back"}
                    </Button>
                </Grid>
                <Grid item xs={10}>
                    <Stepper activeStep={store.current} alternativeLabel>
                        {Children.map(children, createStepLabel)}
                    </Stepper>
                </Grid>
                <Grid item xs={1}>
                    <Button variant="contained"
                            color="primary"
                            onClick={isLastStep ? onFinish : () => store.next()}
                            endIcon={
                                <MdiIcon path={isLastStep ? mdiCheck : mdiArrowRight}
                                         size={1}/>
                            }
                            data-testid={"forward-button"}
                            disabled={!children[store.current].props.isDone?.()}
                    >
                        {isLastStep ? "Finish" : "Next"}
                    </Button>
                </Grid>
            </Grid>
            {children[store.current]}
        </>
    );
});

export default StepperNav;
