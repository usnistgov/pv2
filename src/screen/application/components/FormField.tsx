import {Component} from "react";
import {InputAdornment, TextField} from "@material-ui/core";
import {InputProps as StandardInputProps} from "@material-ui/core/Input/Input";

interface ComponentProps {
    label: string;
    defaultValue?: string;
    onChange?: (event: { target: { value: string } }) => void;
    startAdornment?: string;
    endAdornment?: string;
}

class FormField extends Component<ComponentProps, any> {
    render() {
        return (
            <TextField fullWidth
                       label={this.props.label}
                       defaultValue={this.props.defaultValue}
                       variant={"filled"}
                       onChange={this.props.onChange}
                       InputProps={this.getInputProps()}
            />
        );
    }

    getInputProps(): StandardInputProps {
        let inputProps: StandardInputProps = {};

        if (this.props.startAdornment !== undefined) {
            inputProps['startAdornment'] = <InputAdornment position={"start"}>this.props.startAdornment</InputAdornment>
        }
        if (this.props.endAdornment !== undefined) {
            inputProps['endAdornment'] = <InputAdornment position={"end"}>this.props.endAdornment</InputAdornment>
        }

        return inputProps;
    }
}

export default FormField;