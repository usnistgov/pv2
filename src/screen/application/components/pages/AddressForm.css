.form-container {
  padding-left: 64px;
  padding-right: 64px;
}

.map {
  padding: 0;
  height: 100%;
}
.map iframe {
  width: 100%;
  height: 100%;
}

/*# sourceMappingURL=AddressForm.css.map */
