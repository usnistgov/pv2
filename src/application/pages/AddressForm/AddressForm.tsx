import React, {useContext} from "react";

// Library Imports
import {Grid, Paper} from "@material-ui/core";
import * as Yup from "yup";
import {observer} from "mobx-react-lite";
import {action} from "mobx";

// User Imports
import MaterialHeader from "../../../components/MaterialHeader/MaterialHeader";
import {Store} from "../../ApplicationStore";
import {
    ADDRESS_LABEL,
    ADDRESS_TOOLTIP,
    CITY_LABEL,
    CITY_TOOLTIP,
    STATE_LABEL,
    STATE_TOOLTIP,
    ZIPCODE_LABEL,
    ZIPCODE_TOOLTIP
} from "../../../Strings";
import ValidatedTextField from "../../../components/ValidatedTextField";
import Info from "../../../components/Info/Info";
import ResetButton from "../../../components/ResetButton/ResetButton";

// Stylesheets
import "./AddressForm.sass"

/*
 * The AddressForm component is the first page of the application form that lets user fill
 * in their address and displays an embedded iframe of Google Maps with the inputted location.
 */
const AddressForm = observer(() => {
    const store = useContext(Store).addressFormStore;

    // Google Maps embedded url query
    // TODO replace this with real key
    const wholeMap = !(store.address || store.city || store.state || store.zipcode) ? "&center=39.8097343,-98.5556199&zoom=3" : "";
    const query = `https://www.google.com/maps/embed/v1/place?key=AIzaSyDGXJiK0XkDxlx2loXvonuX6BJOIYpd0Lg&q=${store.address}, ${store.city}, ${store.state} ${store.zipcode}${wholeMap}`;

    return (
        <>
            <MaterialHeader text={"Address"} right={<ResetButton onClick={() => store.reset()}/>}/>
            <div className={"address-form-text"}>
                Insert address for home on which a solar PV system may be installed.
                Only the ZIP Code is necessary for [PV]<sup>2</sup> to calculate results.
            </div>
            <Grid className={"address-form-container"} container justifyContent={"center"} spacing={8}>
                <Grid item xs={6}>
                    <Grid container spacing={2}>
                        <Grid item xs={12}>
                            <Info tooltip={ADDRESS_TOOLTIP}>
                                <ValidatedTextField fullWidth
                                                    label={ADDRESS_LABEL}
                                                    value={store.address}
                                                    variant={"filled"}
                                                    schema={Yup.string()}
                                                    onValidate={action((value) => store.address = value)}/>
                            </Info>
                        </Grid>
                        <Grid item xs={12}>
                            <Info tooltip={CITY_TOOLTIP}>
                                <ValidatedTextField fullWidth
                                                    label={CITY_LABEL}
                                                    value={store.city}
                                                    variant={"filled"}
                                                    schema={Yup.string()}
                                                    onValidate={action((value) => store.city = value)}/>
                            </Info>
                        </Grid>
                        <Grid item xs={12}>
                            <Info tooltip={STATE_TOOLTIP}>
                                <ValidatedTextField fullWidth
                                                    label={STATE_LABEL}
                                                    value={store.state}
                                                    variant={"filled"}
                                                    schema={Yup.string()}
                                                    onValidate={action((value) => store.state = value)}/>
                            </Info>
                        </Grid>
                        <Grid item xs={12}>
                            <Info tooltip={ZIPCODE_TOOLTIP}>
                                <ValidatedTextField fullWidth
                                                    required
                                                    label={ZIPCODE_LABEL}
                                                    value={store.zipcode}
                                                    variant={"filled"}
                                                    schema={Yup.string().required()}
                                                    onValidate={action((value) => store.zipcode = value)}/>
                            </Info>
                        </Grid>
                    </Grid>
                </Grid>
                <Grid item xs={6}>
                    <Paper className={"map"} elevation={3}>
                        <iframe title={"Google Address Map"} frameBorder="0" src={query} allowFullScreen/>
                    </Paper>
                </Grid>
            </Grid>
        </>
    );
})

export default AddressForm;
