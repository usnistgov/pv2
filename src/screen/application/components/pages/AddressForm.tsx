import {Component} from "react";
import {Box, Grid, Paper, TextField} from "@material-ui/core";
import MaterialHeader from "../MaterialHeader";
import "./AddressForm.css"
import FormField from "../FormField";
import {RootState} from "../../ApplicationStore";
import {connect} from "react-redux";
import {addressSlice, citySlice, stateSlice, zipcodeSlice} from "./AddressFormReducer"
import * as Yup from 'yup';

interface DispatchProps {
    setAddress: (s: string) => void;
    setCity: (s: string) => void;
    setState: (s: string) => void;
    setZipcode: (s: string) => void;
}

interface StateProps {
    address: string;
    city: string;
    state: string;
    zipcode: string;
}

interface ComponentState {
    errors: {
        address: string,
        city: string,
        state: string,
        zipcode: string,
    }
}

type Props = DispatchProps & StateProps;

class AddressForm extends Component<Props, ComponentState> {
    constructor(props: Props){
        super(props);

        this.state = {
            errors: {
                address: "",
                city: "",
                state: "",
                zipcode: "",
            }
        }
    }

    render() {
        let schema = Yup.object({
            address: Yup.string()
                .required(),
            city: Yup.string()
                .required(),
            state: Yup.string()
                .required(),
            zipcode: Yup.string()
                .required(),
        });

        return (
            <Box>
                <MaterialHeader text={"Address"}/>
                <Grid className={"form-container"} container justify={"center"} spacing={8}>
                    <Grid item xs={6}>
                        <Grid container spacing={2}>
                            <Grid item xs={12}>
                                <TextField fullWidth
                                           error={this.state.errors.address !== ""}
                                           helperText={this.state.errors.address}
                                           label={"Address"}
                                           defaultValue={this.props.address}
                                           variant={"filled"}
                                           onChange={(event) => {
                                               if(event.target.value === ""){
                                                   this.setState((state) => {
                                                       state.errors.address = "required";
                                                       return state;
                                                   })
                                               } else {
                                                   this.setState((state) => {
                                                       state.errors.address = "";
                                                       return state;
                                                   })
                                               }
                                           }}
                                />
                            </Grid>
                            {["Address", "City", "State", "Zipcode"].map(this.createFormField)}
                        </Grid>
                    </Grid>
                    <Grid item xs={6}>
                        <Paper className={"map"} elevation={3}>
                            <iframe
                                title={"Google Address Map"}
                                frameBorder="0"
                                src={`https://www.google.com/maps/embed/v1/place?key=AIzaSyDGXJiK0XkDxlx2loXvonuX6BJOIYpd0Lg
                            &q=${this.props.address}, ${this.props.city}, ${this.props.state} ${this.props.zipcode}`}
                                allowFullScreen/>
                        </Paper>
                    </Grid>
                </Grid>
            </Box>
        );
    }

    lift(func: (s: string) => void): (event: { target: { value: string } }) => void {
        return (event) => func(event.target.value);
    }

    createFormField = (label: string) => (
        <Grid item xs={12}>
            <FormField label={label}
                       defaultValue={this.props[label.toLowerCase() as keyof StateProps]}
                       onChange={this.lift(this.props[`set${label}` as keyof DispatchProps])}/>
        </Grid>
    );
}

/*
 * Redux mappings and export
 */
function mapStateToProps(state: RootState): StateProps {
    return {
        address: state.address,
        city: state.city,
        state: state.state,
        zipcode: state.zipcode,
    }
}

const mapDispatchToProps = {
    setAddress: (s: string) => addressSlice.actions.set(s),
    setCity: (s: string) => citySlice.actions.set(s),
    setState: (s: string) => stateSlice.actions.set(s),
    setZipcode: (s: string) => zipcodeSlice.actions.set(s),
}

export default connect(mapStateToProps, mapDispatchToProps)(AddressForm);