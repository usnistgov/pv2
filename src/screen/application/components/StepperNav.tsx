import {Children, Component, ReactElement, ReactNode} from 'react';

import {Box, Button, Grid, Step, StepLabel, Stepper} from "@material-ui/core";
import {Icon as MdiIcon} from "@mdi/react"
import {mdiArrowLeft, mdiArrowRight} from "@mdi/js";
import StepperPage from "./StepperPage";
import {RootState} from "../ApplicationStore";
import {connect} from "react-redux";
import {increment, decrement} from "./StepperNavReducer";
import "./StepperNav.css"

/*
 * Property interfaces
 */
interface ComponentProps {
    children: ReactElement[];
}

interface DispatchProps {
    increment: () => void;
    decrement: () => void;
}

interface StateProps {
    activeStep: number;
}

/*
 * The StepperNav Component creates a Stepper navigation bar with back and next buttons and contains a list
 * of pages that are displayed according to the current active step.
 */
class StepperNav extends Component<ComponentProps & DispatchProps & StateProps> {
    render(): ReactNode {
        return (
            <Box>
                <Grid className={"grid-container"} container justify={'center'} alignItems={'center'} spacing={0}>
                    <Grid item xs={1}>
                        <Button
                            disabled={this.props.activeStep === 0}
                            onClick={this.props.decrement}
                            startIcon={
                                <MdiIcon path={mdiArrowLeft}
                                         size={1}/>
                            }
                        >
                            Back
                        </Button>
                    </Grid>
                    <Grid item xs={10}>
                        <Stepper activeStep={this.props.activeStep} alternativeLabel>
                            {Children.map(this.props.children, this.createStepLabel)}
                        </Stepper>
                    </Grid>
                    <Grid item xs={1}>
                        <Button variant="contained"
                                color="primary"
                                disabled={this.props.activeStep === this.props.children.length - 1}
                                onClick={this.props.increment}
                                endIcon={
                                    <MdiIcon path={mdiArrowRight}
                                             size={1}/>
                                }
                        >
                            Next
                        </Button>
                    </Grid>
                </Grid>
                {this.props.children[this.props.activeStep]}
            </Box>
        );
    };

    createStepLabel(child: any): ReactNode {
        let page = child as StepperPage;

        return (
            <Step key={page.props.label}>
                <StepLabel>{page.props.label}</StepLabel>
            </Step>
        );
    }
}

/*
 * Redux mappings and export
 */
function mapStateToProps(state: RootState): StateProps {
    return {
        activeStep: state.activeStep
    }
}

export default connect(mapStateToProps, {increment, decrement})(StepperNav);