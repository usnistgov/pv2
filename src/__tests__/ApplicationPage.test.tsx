import React from 'react'
import ApplicationPage from '../screen/application/Application'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import '../setupTests'
import { mount } from 'enzyme'
import StepperNav from '../screen/application/components/StepperNav/StepperNav'
import AddressForm from '../screen/application/pages/AddressForm/AddressForm'
import FormField from '../screen/application/components/FormField/FormField'

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom') as any,
  __esModule: true,
  useHistory: () => ({
    push: jest.fn(),
  }),
}));

describe("<StepperNav />", () => {
  test('StepperNav containes same as snapshot', async () => {
    const wrapper = mount(<StepperNav><div/><div/></StepperNav>)
    expect(wrapper).toMatchSnapshot()
  })

  test('StepperNav containes two buttons', async () => {
    const wrapper = mount(<StepperNav><div/><div/></StepperNav>)
    expect(wrapper.find("button")).toHaveLength(2)
  })
})

describe("<ApplicationPage />", () => {
  test('application page same as snapshot', async () => {
    const wrapper = renderApplicationPage()
    expect(wrapper).toMatchSnapshot()
  })

  test('application page should have stepper nav', async () => {
    const wrapper = renderApplicationPage()
    expect(wrapper.find("StepperNav")).toHaveLength(1)
  })

  test('should be on address page', async () => {
    const wrapper = renderApplicationPage()
    expect(wrapper.find(AddressForm)).toHaveLength(1)
  })
})

describe("<AddressForm />", () => {
  test('Address Form same as snapshot', async () => {
    const wrapper = renderAddressForm()
    expect(wrapper).toMatchSnapshot()
  })

  test('Address Form has 4 inputs', async () => {
    const wrapper = renderAddressForm()
    const formFields = wrapper.find(FormField)
    expect(formFields).toHaveLength(4)
  })

  test('Address Form first input is Address', async () => {
    const wrapper = renderAddressForm()
    const field = wrapper.find(FormField).at(0)
    expect(field.prop('label')).toEqual('Address')
    expect(field.prop('required')).toEqual(true)
  })

  test('Address Form second input is City', async () => {
    const wrapper = renderAddressForm()
    const field = wrapper.find(FormField).at(1)
    expect(field.prop('label')).toEqual('City')
    expect(field.prop('required')).toEqual(true)
  })

  test('Address Form third input is State', async () => {
    const wrapper = renderAddressForm()
    const field = wrapper.find(FormField).at(2)
    expect(field.prop('label')).toEqual('State')
    expect(field.prop('required')).toEqual(true)
  })

  test('Address Form fourth input is Zipcode', async () => {
    const wrapper = renderAddressForm()
    const field = wrapper.find(FormField).at(3)
    expect(field.prop('label')).toEqual('Zipcode')
    expect(field.prop('required')).toEqual(true)
  })
})

function renderAddressForm() {
  const mockStore = configureStore()
  const store: any = mockStore({ placeholder: 0 })
  store.injectReducer = () => { return undefined }
  return mount(<Provider store={store}><AddressForm /></Provider>)
}

function renderApplicationPage() {
  const mockStore = configureStore()
  const store: any = mockStore({ placeholder: 0 })
  store.injectReducer = () => { return undefined }
  return mount(<Provider store={store}><ApplicationPage /></Provider>)
}
