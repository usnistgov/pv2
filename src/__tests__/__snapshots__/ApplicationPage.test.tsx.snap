// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddressForm /> Address Form same as snapshot 1`] = `ReactWrapper {}`;

exports[`<ApplicationPage /> application page same as snapshot 1`] = `ReactWrapper {}`;

exports[`<ElectricalRateForm /> Electrical Rate Form same as snapshot 1`] = `ReactWrapper {}`;

exports[`<StepperNav /> StepperNav containes same as snapshot 1`] = `ReactWrapper {}`;
